### DISCUSSION TITLE
Understanding identify -verbose output

### DISCUSSION THREAD

**Discussion by dmill-bz** [ImageMagick/ImageMagick#3697] [Help]
Hey guys, I ran an `identify -verbose` command on an 8bit png I had and there are a couple of points I would like to clarify (not sure I understand them). 
This image is basically an object (some of the object's pixels are semi transparent) on a transparent background that turns white during the render.

Bellow is the output of `identify -verbose`:
```
Image:
  Filename: /path/to/0015651fa9a4223db655e7f8a67ee4b8.png
  Format: PNG (Portable Network Graphics)
  Mime type: image/png
  Class: DirectClass
  Geometry: 768x693+0+0
  Units: Undefined
  Colorspace: sRGB
  Type: TrueColorAlpha
  Base type: Undefined
  Endianness: Undefined
  Depth: 8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
    Alpha: 8-bit
  Channel statistics:
    Pixels: 532224
    Red:
      min: 0  (0)
      max: 255 (1)
      mean: 12.4236 (0.04872)
      median: 0 (0)
      standard deviation: 32.468 (0.127325)
      kurtosis: 15.8392
      skewness: 3.69072
      entropy: 0.294974
    Green:
      min: 0  (0)
      max: 255 (1)
      mean: 12.037 (0.0472038)
      median: 0 (0)
      standard deviation: 31.9865 (0.125437)
      kurtosis: 16.4471
      skewness: 3.76405
      entropy: 0.294825
    Blue:
      min: 0  (0)
      max: 255 (1)
      mean: 12.1919 (0.0478113)
      median: 0 (0)
      standard deviation: 32.1921 (0.126243)
      kurtosis: 16.1317
      skewness: 3.72915
      entropy: 0.295062
    Alpha:
      min: 0  (0)
      max: 255 (1)
      mean: 62.8626 (0.24652)
      median: 0 (0)
      standard deviation: 107.428 (0.421286)
      kurtosis: -0.588408
      skewness: 1.16792
      entropy: 0.177715
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 24.8788 (0.0975638)
      median: 0 (0)
      standard deviation: 51.0187 (0.200073)
      kurtosis: 7.03017
      skewness: 2.87411
      entropy: 0.265644
  Alpha: none   #00000000
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 768x693+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Zip
  Orientation: Undefined
  Properties:
    date:create: 2021-05-18T22:22:33+00:00
    date:modify: 2021-05-18T22:08:45+00:00
    png:IHDR.bit-depth-orig: 8
    png:IHDR.bit_depth: 8
    png:IHDR.color-type-orig: 6
    png:IHDR.color_type: 6 (RGBA)
    png:IHDR.interlace_method: 0 (Not interlaced)
    png:IHDR.width,height: 768, 693
    png:sRGB: intent=0 (Perceptual Intent)
    signature: 1160c6a249bd9201e30f16f84bad720ad470168fe3876d581443ebe00307f3d4
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 218964B
  Number pixels: 532224
  Pixels per second: 31.6453MP
  User time: 0.030u
  Elapsed time: 0:01.016
  Version: ImageMagick 7.0.11-12 Q16 x86_64 2021-05-09 https://imagemagick.org
```
So, this looks like an rgba image (has an alpha channel). 
The things I'm not too sure about are:
- I'm guessing that the `Alpha color: none #000000` is normal and this would only change if I were trying to get transparency by using an actual color?
- `Background color: white` : Dictates which color will replace the transparent pixels??
- `Matte color: grey74` What is this? Does the background color above only replace fully transparent pixels? Is this some kind of mask data? Is this needed in any way? I'm a little confused about this line and can't seem to find any documentation to help?

**Comment by snibgo** [comment]
I don't know what "Alpha color" is used for, if anything.

The "Matte color" is used by "-frame" and "-distort".

The "Background color" is used in many operations such as "-rotate". You can set it with the "-background" option.

**Reply by Jmuccigr** [reply]
To me it seems odd that `identify` would include information that is not stored in the file. That is, png can hold background info, but jpeg and tiff cannot, yet background color is reported anyway.

**Reply by dmill-bz** [reply]
I've figured out that alpha color is the color that trasnparent pixels take when we set alpha to be opaque. 
A basic alpha color will be rgba(x,y,z,0) < since it's meant to be transparent the color is mostly irrelevant (you can see that the a = 0) But in some cases (like converting your PNG to JPEG without `-flatten`) that opacity will be set to 100% and all the fully transparent pixels will take on the color rgb(x,y,z). All the semi-transparent pixels however will get turned to black. I'm not sure why or where this pixel color comes from. Maybe it's just `null` data (`#000000`) 

When it comes to Background color. This is as it sounds. For PNGs think of this as the bottom layer of the image. It's a monochrome color. If you convert a PNG to JPEG with `-flatten` then the transparent image will be flattened onto the background color (instead of the alpha color becoming opaque as mentioned above)

Matte color is used for frames or to fill in missing parts of the image on some manipulations (apparently)

**Comment by snibgo** [comment]
"identify" shows metadata about the image(s) as they are currently recorded in memory. Some of this metadata comes from the file, but much does not.

If you care only about metadata stored in files, use "exiftool".

**Reply by Jmuccigr** [reply]
Ah, good to know.
